### PR TITLE
Add chat history feature

### PR DESCRIPTION
**Pull request by fujita-h**
This commit adds the following new files:
- HistoryItem: A new component for displaying chat history items.
- HistoryPanel: A new component for displaying the chat history panel.
- HistoryButton: A new component for opening the chat history.
- HistoryProviders: A new module containing history provider classes.

These changes include functionality to support future enhancements to the chat history feature.

The chat history feature is only enabled when you explicitly set the environment variable `USE_CHAT_HISTORY`.

The history function uses the browser's IndexedDB. I understand that there is also an approach to use CosmosDB, but this requires enabling authentication to identify users, so I decided to start with IndexedDB, which does not depend on authentication.

To make it easier to implement other data stores, I made it easier to implement additional data stores with `HistoryProviders`.

Related: #977 #1863 #259 

## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

```
[ ] Yes
[x] No
```

## Does this require changes to learn.microsoft.com docs?

This repository is referenced by [this tutorial](https://learn.microsoft.com/azure/developer/python/get-started-app-chat-template)
which includes deployment, settings and usage instructions. If text or screenshot need to change in the tutorial,
check the box below and notify the tutorial author. A Microsoft employee can do this for you if you're an external contributor.

```
[ ] Yes
[x] No
```

## Type of change

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Code quality checklist

See [CONTRIBUTING.md](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/CONTRIBUTING.md#submit-pr) for more details.

- [x] The current tests all pass (`python -m pytest`).
- [ ] I added tests that prove my fix is effective or that my feature works
- [x] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [x] I ran `python -m mypy` to check for type errors
- [x] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

### CODE REVIEW

**Comment by fujita-h** [comment]
@microsoft-github-policy-service agree

**Comment by pamelafox** [comment]
Ooo, very interesting approach with the IndexedDB! It sounds like you've designed it very thoughtfully, I'll take a look soon.

File: app/frontend/src/pages/chat/Chat.tsx
Comment (pamelafox) [inline app/frontend/src/pages/chat/Chat.tsx:207]: Do you think we'd also use this randomUUID for apps with logged in users, or would we use Entra OIDs for them? Am wondering if it'd be better to have a helper function to generate this ID.
And if the history storage was in CosmosDB, would we still generate the UUID on the client? Want to make sure this is forward-compatible if possible.

File: app/frontend/package.json
Comment (pamelafox) [inline app/frontend/package.json:25]: This seems like a reasonable addition according to packagephobia and snyk:

https://packagephobia.com/result?p=idb
https://snyk.io/advisor/npm-package/idb

File: app/backend/config.py
Comment (pamelafox) [inline app/backend/config.py:25]: Given that a "chat history" feature means different things to different people, I think it might help for the config and env variable to be more specific.

For example, for the env var:

USE_CHAT_HISTORY_BROWSER

In future, we might have:
USE_CHAT_HISTORY_COSMOS

That'd be similar to how we did speech output.

File: app/frontend/src/locales/es/translation.json
Comment (pamelafox) [inline app/frontend/src/locales/es/translation.json:10]: Did you use AI for the translations? i.e. do we need a human to review them?

File: app/frontend/src/api/models.ts
Comment (pamelafox) [inline app/frontend/src/api/models.ts:92]:
```suggestion
    showChatHistoryBrowser: boolean;
```

Comment (pamelafox) [reply app/frontend/src/api/models.ts:92]: Per my other comment, I think we should be clear about what's enabled.

File: app/frontend/src/pages/chat/Chat.tsx
Comment (fujita-h) [reply app/frontend/src/pages/chat/Chat.tsx:207]: Thank you for pointing that out.
You're right, it seems we should think about this carefully.

We need to assign an ID to each "series of chats" to identify them, regardless of whether the user is logged in or not. We'll call this `sessionId`.

## In the case of IndexedDB
- On the client side (React), `sessionId` does not necessarily have to be a UUID, as long as it is a unique key within the client side.
- There is also no need to exchange `sessionId` with the server side using `session_state` (the current implementation of storing `sessionId` in `session_state` is in anticipation of expansion to CosmosDB).

## In the case of CosmosDB
- It is expected that the design will set the user's Entra OID as the partition key and `sessionId` as the unique key.
- In CosmosDB, `conversationId` does not need to be a UUID, as long as it is unique within the partition.
- As mentioned above, `sessionId` does not need to be a UUID on the client side (React).
- `sessionId` is a key in CosmosDB. Therefore, it is preferable to generate it on the server side rather than on the client side.
- As the name `session_state` suggests, it is natural that the `sessionId` stored here is generated on the server side.
- When chat messages are sent and received, the `sessionId` needs to be exchanged between the server side and the client side. `session_state` can be used for this.

## Fix Plan
The plan is to summarize these points and fix them in the following plan.
- If the request from the client contains `session_state`, it will be used as `sessionId`. If not, a new one will be generated on the server and added to `session_state`.
- `sessionId` does not necessarily need to be UUID, but unless there is a specific reason, there is no major problem with UUID, so we will basically generate it with `uuid.uuid4()`.
- Considering the possibility that UUID may be inconvenient when other storage methods are added in the future, we will create a helper function that generates `sessionId` according to the storage method used.

File: app/frontend/src/locales/es/translation.json
Comment (pamelafox) [reply app/frontend/src/locales/es/translation.json:10]: Okay, will do!

**Comment by fujita-h** [comment]
I have corrected the points you pointed out, so please review again.
The environment variable to enable this feature has been changed to `USE_CHAT_HISTORY_BROWSER`.

**Comment by fujita-h** [comment]
Thank you for pointing that out. I had forgotten about bicep.
`ENABLE_LANGUAGE_PICKER` and `USE_SPEECH_INPUT_BROWSER` are outside the scope of this PR, so I have omitted them.

File: app/frontend/src/locales/fr/translation.json
Comment (manekinekko) [inline app/frontend/src/locales/fr/translation.json:10]:
```suggestion
        "chatHistory": "Historique du chat",
```

File: app/frontend/src/locales/fr/translation.json
Comment (manekinekko) [inline app/frontend/src/locales/fr/translation.json:11]:
```suggestion
        "openChatHistory": "Ouvrir l'historique du chat",
```

File: app/frontend/src/locales/fr/translation.json
Comment (manekinekko) [inline app/frontend/src/locales/fr/translation.json:14]:
```suggestion
        "deleteModalDescription": "Cette action est irréversible. Supprimer l'historique de ce chat ?",
```

File: app/frontend/src/locales/fr/translation.json
Comment (manekinekko) [inline app/frontend/src/locales/fr/translation.json:12]:
```suggestion
        "noHistory": "Pas d'historique dans ce chat",
```

File: app/frontend/src/locales/fr/translation.json
Comment (manekinekko) [inline app/frontend/src/locales/fr/translation.json:13]:
```suggestion
        "deleteModalTitle": "Supprimer l'historique du chat",
```

File: app/frontend/src/locales/fr/translation.json
Comment (pamelafox) [reply app/frontend/src/locales/fr/translation.json:14]: Merci, Wassim!

File: app/frontend/src/locales/es/translation.json
Comment (pamelafox) [reply app/frontend/src/locales/es/translation.json:10]: I've asked a colleague to review the es translations. Wassim has reviewed fr.

**Review by pamelafox — APPROVED** [review]
I'm approving the change, but will await a review of the es translations and will check with @mattgotteiner before merging.

**Review by mattgotteiner — APPROVED** [review]
approved pending translation review

File: app/frontend/src/locales/es/translation.json
Comment (miguelmsft) [reply app/frontend/src/locales/es/translation.json:10]: I approve of the Spanish translation 👍

Comment (pamelafox) [reply app/frontend/src/locales/es/translation.json:10]: Thanks miguel! Resolving this thread.

**Review by pamelafox — APPROVED** [review]

**Comment by ibrahimnasir0** [comment]
@fujita-h Where we will be able to see this chat history  ?
Can you guide on that

**Comment by fujita-h** [comment]
@ibrahimnasir0 Just set the environment variables as follows:
`azd env set USE_CHAT_HISTORY_BROWSER true`

**Comment by ibrahimnasir0** [comment]
@fujita-h Thanks for the help! By the way, would you mind sharing your LinkedIn profile? I would love to connect and discuss the implementation further

**Comment by ibrahimnasir0** [comment]
Hi @fujita-h 
I'm working on integrating Cosmos DB with this updated project to save chat history along with the user OID (since login and Entra ID are enabled). Could you provide some guidance on how to proceed with this?

**Pull request merged**
